Serva w1 browser

Wandering-Souls opened the pull request:
Added w1 browser exercises!

Erlemorgaine commented:
Hi Serva, 

Overall your work looks good! I'm impressed, since I know the exercises aren't very easy. It will seem like I have a lot of comments, but don't worry about that, most of them are really only meant to help you write more readable code :)

### Exercise 1

* The exercise looks all good, you made only one small mistake: I think you wanted to set the image alt text to the title of the book, but actually this code causes the alt text to be literally '&{books.title}' (if you open the html file in your browser and inspect the image element, you'll also see what I mean). Can you fix this?

```
image.alt = `&{books.title}`;
```

(Another tip here: you actually don't even need the template literal here, you can just assign `books.title` directly to the `image.alt`). 

Two small things you can do to improve your code:

* You placed the variable `coverImage` inside the `.forEach` loop. So now, every time the `.forEach` loop runs, the variable `coverImage` is made again. However, this isn't necessary, since the `coverImage` has the same value for every iteration. So, we can just place the variable outside of the `.forEach` loop, and reuse it for each iteration. 

* You're using the ternary operator to make a book red or green. Can you think of a way to rewrite (refactor) this code so that it's shorter? (Hint: in the refactored version you only need to write `bookListItems.style.background` once.)

```
book.alreadyRead
      ? (bookListItems.style.background = 'green')
      : (bookListItems.style.background = 'red');
```

Then, some cases in which you can improve your variable names:

* The variable `bookListItems` actually only contains one item, can you rename it so that it's singular?
* The opposite: the variable `coverImage` actually contains multiple images, can you rename it so that it's plural?


### Exercise 2

* If I open the index.html file in my browser, the code looks almost correct. But there is one thing missing. Can you check what it is and correct it? (Hint: the thing is missing because you made a typo in the code, in an important place :)). 

* You added a font family, like this: `document.body.style.fontFamily = 'Arial, sans-serif'`. Now you are still in the phase where it's completely okay to write 'inline styles' like these. But once you start working as a developer, it's better to use as little inline styles as possible, and always try to use a css file with classes. One reason for this is that it keeps your code more organised. So with that in mind, instead of setting the font family like this, could you use the `list-item` class to set the font family directly in the CSS file?
 

### Exercise 3

Nothing to add here, all good!


### Exercise 4

* Half of the exercise looks good, but I think you miss the second half of the exercise: I think the time also needs to be displayed on your web page, in html :). The README says: "Verify that the current time is displayed correctly on both the web page and the browser's console."

* I think the idea of the exercise is that you actually use `setInterval` _inside_  your function. That would also make sense, because then you will be able to call your function less times manually (see my next point). Can you rewrite your code to use `setInterval` inside the `addCurrentTime` function?

* Now you manually use your function 3 times at the end of your file. But actually, you only need to use it once. Can you rewrite your code so that you'll only need to use it once? (For this you also need to move `setInterval` into the function, as mentioned before). Basically, once you've done that, of these three lines you only need to keep one line:

```
addCurrentTime();
window.addEventListener('load', addCurrentTime);
setInterval(addCurrentTime, 1000);
```


### Exercise 5

- I really like your solution! You call the function `catWalk` in a recursive way, that's really good. There is one problem: when I open the index.html file, my cat is not dancing. The reason for this is this line:

```
moveToRight === (window.innerWidth - catImage.width) / 2
```

Here you say that `moveToRight` has to be exactly the same value as the value of `(window.innerWidth - catImage.width) / 2`. But, the value of `moveToRight` will always be a value of tens: it will be 10, 20, 30, 40, 50, etc. 
However, your browser window will not always be exactly a value of tens. For example, it can happen that your window is 1010px wide. So half of your screen will be 505px. 
In that case `moveToRight` will never `(window.innerWidth - catImage.width) / 2`, since `moveToRight` will move directly from 500 to 510. 

Can you rewrite this part of your code in such a way that it still works if the value `(window.innerWidth - catImage.width) / 2` is not cleanly divisible by 10?

Wandering-Souls commented:
Hello Erle, 
Thank you very much for correction and all the great tips you shared, I learn a lot from your feedback. I did most of these exercises at midnight so I already expected to have some issues! 

About inline styling I do agree with you and actually in case of ex-1 I did it in CSS first but then I commented out and wanted to try it in JS file just for practicing what I have learned in this week. 

Ex1. 

- I have changed my `image.alt`.
- I've put my coverImages outside the loop.
- The name of coverImage and bookListItems are modified.
- I've changed my ternary condition to make it shorter.

Ex-2
- I found the bug in my favorite food statement.
- Style.font-family is transferred to CSS file.

Wandering-Souls commented:
Ex-4

I don't know why I haven't done the second part! I could see the time in my terminal so I thought I'm done.

- Now, you can see the time in the browser.
- `setInterval` has moved inside the function. 
- I've called the function only once : `window.addEventListener('load', addCurrentTime);`

Wandering-Souls commented:
Ex-5
My cat is dancing on my screen so based on what you explained my screen width - cat.width is a value of 10. 

- I changed my code and checked the cat on screen with different width and it's responsive/dancing in the middle.  

**_Thanks again for the feedback, I really appreciate your help!_**

Erlemorgaine commented:
Hi Serva, thank you for the changes! 

**Exercise 1 and 5** are all good now. 
(I have another small feedback about exercise 5, but since exercise 5 is quite complicated and since I have this feedback for everyone, we'll discuss it during the feedback hour this afternoon). 

I have to apologise for my feedback in **Exercise 2**, because it seems it actually broke one test! Judging by the tests, the HYF team actually wanted you to put the fontFamily as an inline style instead of in the .css file. So I'm afraid to have to ask you to remove `font-family: 'Arial, sans-serif';` from the css file and re-add this line in the Javascript file:

```
bodyEl.style.fontFamily = 'Arial, sans-serif';
```

Sorry again! My bad :)


Then, about **Exercise 4**: I can see the time in the browser now, so that's good! Unfortunately, some other things got broken in the process. You'll see that one test fails. And also, when you'll open your Devtools console to check the logs there, you'll see what the problem is. 

There is also a problem with this line: `setInterval(addCurrentTime, 1000);`. On first look it seems to work, but when you fix your console.logs, you will probably notice something strange with the logs. But, I think it's more easy if I explain this in the feedback hour. 


One general tip: whenever you've made changes to your code, it's best to always rerun the tests, and check your html file in the browser (with the DevTools console open). In this way, you'll see it directly when things are broken :).

Wandering-Souls commented:
Hi Erle, 
Thank your for the feedback again. 
Regarding ex-4, last time I put my `set interval` inside the function based on your comment but now I moved it outside the function again and I don't see the error anymore in console.log, it also passed all the tests. But to be honest, this is not clear to me, what's the difference between putting set interval inside or outside a function. Let's discuss that in our session today.

Wandering-Souls commented:
Ex-2.
I changed the font-family and now it passes all the tests.
